light-clients/ics10-grandpa-cw/src/contract.rs

```diff
@@ -0,0 +1,150 @@
+}
+
+impl grandpa_light_client_primitives::HostFunctions for HostFunctions {
+	type Header = RelayChainHeader;
+
+	fn ed25519_verify(
+		_sig: &sp_core::ed25519::Signature,
+		_msg: &[u8],
+		_pub_key: &sp_core::ed25519::Public,
+	) -> bool {
+		todo!()
+	}
+
+	fn insert_relay_header_hashes(_headers: &[<Self::Header as Header>::Hash]) {
+		todo!()
+	}
+
+	fn contains_relay_header_hash(_hash: <Self::Header as Header>::Hash) -> bool {
+		todo!()
+	}
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+	_deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	_msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+	Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+	deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+	let deps = Rc::new(RefCell::new(deps));
+	let client = GrandpaClient::<HostFunctions>::default();
+	let ctx = Context::<HostFunctions>::new(deps);
+	match msg {
+		ExecuteMsg::ValidateMsg(_) => todo!(),
+		ExecuteMsg::StatusMsg(_) => todo!(),
+		ExecuteMsg::ExportedMetadataMsg(_) => todo!(),
+		ExecuteMsg::ZeroCustomFieldsMsg(_) => todo!(),
+		ExecuteMsg::GetTimestampAtHeightMsg(_) => todo!(),
+		ExecuteMsg::InitializeMsg(_) => todo!(),
+		ExecuteMsg::VerifyMembershipMsg(msg) => {
+			let msg = VerifyMembershipMsg::try_from(msg)?;
+			verify_membership::<BlakeTwo256, _>(
+				&msg.prefix,
+				&msg.proof,
+				&msg.root,
+				msg.path,
+				msg.value,
+			)
+			.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+		},
+		ExecuteMsg::VerifyNonMembershipMsg(msg) => {
+			let msg = VerifyNonMembershipMsg::try_from(msg)?;
+			verify_non_membership::<BlakeTwo256, _>(&msg.prefix, &msg.proof, &msg.root, msg.path)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+		},
+		ExecuteMsg::VerifyClientMessage(msg) => {
+			let msg = VerifyClientMessage::try_from(msg)?;
+			client
+				.verify_client_message(&ctx, msg.client_id, msg.client_state, msg.client_message)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+		},
+		ExecuteMsg::CheckForMisbehaviourMsg(msg) => {
+			let msg = CheckForMisbehaviourMsg::try_from(msg)?;
+			client
+				.check_for_misbehaviour(&ctx, msg.client_id, msg.client_state, msg.client_message)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+		},
+		ExecuteMsg::UpdateStateOnMisbehaviourMsg(msg) => {
+			let msg = UpdateStateOnMisbehaviourMsg::try_from(msg)?;
+			client
+				.update_state_on_misbehaviour(msg.client_state, msg.client_message)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+		},
+		ExecuteMsg::UpdateStateMsg(msg) => {
+			let msg = UpdateStateMsg::try_from(msg)?;
+			client
+				.update_state(&ctx, msg.client_id, msg.client_state, msg.client_message)
+				.map_err(|e| ContractError::Grandpa(e.to_string()))?;
+		},
+		ExecuteMsg::CheckSubstituteAndUpdateStateMsg(_msg) => {
+			todo!("check substitute and update state")
```

Comment:
I couldn't find a corresponding function in ibc-rs. @seunlanlege do we need it?

Reply:
we don't